liblocation/src/main/java/com/mapbox/android/core/location/MapboxFusedLocationEngineImpl.java

```diff
@@ -0,0 +1,160 @@
+package com.mapbox.android.core.location;
+
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+
+import static com.mapbox.android.core.location.Utils.isBetterLocation;
+
+/**
+ * Mapbox replacement for Google Play Services Fused Location Client
+ * <p>
+ * Note: optimize engine logic for availability
+ */
+class MapboxFusedLocationEngineImpl extends AndroidLocationEngineImpl implements LocationListener {
+  private static final String TAG = "MapboxLocationEngine";
+
+  private LocationEngineCallback<LocationEngineResult> callback;
+  private Location currentBestLocation;
+
+  MapboxFusedLocationEngineImpl(@NonNull Context context) {
+    super(context);
+    this.callback = null;
+  }
+
+  @NonNull
+  @Override
+  LocationListener createListener(final LocationEngineCallback<LocationEngineResult> callback) {
+    this.callback = callback;
+    return this;
+  }
+
+  @Nullable
+  @Override
+  public LocationListener removeLocationListener(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    this.callback = null;
+    return super.removeLocationListener(callback);
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    Location bestLastLocation = getBestLastLocation();
+    if (bestLastLocation != null) {
+      callback.onSuccess(LocationEngineResult.create(bestLastLocation));
+    } else {
+      callback.onFailure(new Exception("Last location unavailable"));
+    }
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull LocationListener listener,
+                                     @Nullable Looper looper) throws SecurityException {
+    super.requestLocationUpdates(request, listener, looper);
```

Comment:
Could you please explain the logic behind possible `locationManager.requestLocationUpdates` here and in `super.` ?
Will we then request updates from locationManager multiple times if `shouldStartNetworkProvider` is true?

Reply:
parent class is responsible for choosing the best provider available at a given moment and starting updates with it, the child class is responsible for figuring out whether these updates (in most cases gps) need to be supplemented with network location updates. if `shouldStartNetworkProvider == true` it is technically dispatches two parallel requests to both network and gps providers and the engine will get fused updates from both.